stablehlo/dialect/TypeInference.cpp

```diff
@@ -910,6 +919,268 @@ LogicalResult inferDotShape(RankedTensorType lhs, RankedTensorType rhs,
+ *     size-of('inserted_window_dims')
+ * P4. size-of('scatter_dims_to_operand_dims') =
+ *         'scatter_indices'['index_vector_dim'] &
+ *     'scatter_dims_to_operand_dims' must be valid indices of 'operand' tensor.
+ */
+LogicalResult validateScatterDimensionNumbers(
+    ShapedType operandType, ArrayRef<int64_t> scatterIndicesShape,
+    ShapedType updateType, bool operandTypeRanked,
+    bool scatterIndicesTypeRanked, bool updatesTypeRanked,
+    ArrayRef<int64_t> updateWindowDims, ArrayRef<int64_t> insertedWindowDims,
+    ArrayRef<int64_t> scatterDimsToOperandDims, int64_t indexVectorDim,
+    Optional<Location> loc) {
+  // P1.
+  if (!llvm::is_sorted(updateWindowDims))
+    return emitOptionalError(loc,
+                             "Expects update_window_dims to be sorted; got: [",
+                             updateWindowDims, "].");
+
+  if (hasDuplicates(updateWindowDims))
+    return emitOptionalError(loc,
+                             "Expects update_window_dims to not repeat; got: [",
+                             updateWindowDims, "].");
+
+  if (updatesTypeRanked) {
+    for (int64_t windowDim : updateWindowDims) {
+      if (windowDim < 0 || windowDim >= updateType.getRank()) {
+        return emitOptionalError(
+            loc,
+            "Expects each element of update_window_dims to be in range "
+            "[0, "
+            "rank-of('updates') i.e. [0, ",
+            updateType.getRank(), "). got: ", windowDim, ".");
+      }
+    }
+  }
+
+  // P2.
+  if (!llvm::is_sorted(insertedWindowDims))
+    return emitOptionalError(
+        loc, "Expects inserted_window_dims to be sorted; got: [",
+        insertedWindowDims, "].");
+
+  if (hasDuplicates(insertedWindowDims))
+    return emitOptionalError(
+        loc, "Expects inserted_window_dims to not repeat; got: [",
+        insertedWindowDims, "].)");
+
+  if (operandTypeRanked) {
+    for (int64_t insertedDim : insertedWindowDims) {
+      if (insertedDim < 0 || insertedDim >= operandType.getRank()) {
+        return emitOptionalError(
+            loc,
+            "Expects each element of inserted_window_dims to be in range "
+            "[0, rank-of('operand') i.e. [0, ",
+            operandType.getRank(), "). got: ", insertedDim, ".");
+      }
+    }
+  }
+
+  // P3.
+  if (operandTypeRanked) {
+    auto windowSize = updateWindowDims.size() + insertedWindowDims.size();
+    if (operandType.getRank() != static_cast<int64_t>(windowSize))
+      return emitOptionalError(loc,
+                               "Expects rank-of operand to match "
+                               "size-of('update_window_dims')  + "
+                               "size-of('inserted_window_dims') i.e. ",
+                               windowSize, " but got ", operandType.getRank(),
+                               ".");
+  }
+
+  // P4.
+  if (scatterIndicesTypeRanked) {
+    if (!hlo::isDynamicDimSize(scatterIndicesShape[indexVectorDim]) &&
```

Comment:
Let's drop all `hlo::` qualifiers from code in TypeInference.cpp, since the file is already in the `hlo::` namespace.